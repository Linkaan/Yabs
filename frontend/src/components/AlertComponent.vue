<template>
  <v-alert
    v-model="show"
    class="onTop"
    :color="color"
    dense
    rounded
    dismissible
    elevation="6"
    prominent
    :type="type"
    :icon="icon"
    transition="scale-transition"
  >
    {{ text }}
  </v-alert>
</template>

<script lang="ts">
import { Ref, computed, watch, defineComponent, ref, SetupContext } from '@vue/composition-api';
import StatusModule from '@/store/modules/StatusModule';
import { StatusMessage } from '@/types';

export default defineComponent({
  name: 'AlertComponent',
  setup(_ : object, ctx: SetupContext): object {
    const show = ref(false);

    const obj = computed(() => StatusModule.allAsArray);

    const color = ref('');
    const type = ref('info');
    const icon: Ref<string | boolean> = ref('mdi-information');
    const text = ref('');

    let timer : ReturnType<typeof setTimeout>;

    // Tries to update the ref properties with data from the latest statusMessage
    watch(obj, (val, prevValue) => {
      if (val && obj.value.length > 0) {
        let temp: StatusMessage;
        try {
          temp = obj.value[obj.value.length - 1];
        } catch{return;} // Invalid data was passed; abort

        color.value = temp.color == undefined ? '' : temp.color;
        text.value = temp.text == undefined ? '' : temp.text;
        type.value = temp.type == undefined ? 'info' : temp.type;
        if (temp.type == undefined) {
          // Valid types are: success, info, warning, and error.
          icon.value = temp.icon == undefined ? 'mdi-information' : temp.icon;
        } else {
          // TODO: Fix prop validation for vuetify icon
          icon.value = true;
        }

        show.value = true;
        if (temp.timeout !== undefined) {
          // Avoids flickering alerts
          clearTimeout(timer);
          timer = setTimeout(() => {
            show.value = false;
          }, temp.timeout);
        }
      }
    });

    return { color, type, icon, text, show };
  },
});
</script>

<style scoped>
.onTop {
  z-index: 99;
  position: fixed;
  top: 66px;
  width: 80vw;
  left: 10vw;
}
</style>
