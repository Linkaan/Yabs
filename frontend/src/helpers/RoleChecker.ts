import UsersModule, { User } from '../store/modules/UsersModule';

export default class RoleChecker {

  private static checkPermission(user: User, required: number): boolean {
    return user && ((user.role & required) == required || (user.role > required));
  }
  
  static isPrincipal(): boolean {
    return this.checkPermission(UsersModule.currentUser, 32);
  }

  static isAdmin(): boolean {
    return this.checkPermission(UsersModule.currentUser, 16);
  }

  static isCaretaker(): boolean {
    return this.checkPermission(UsersModule.currentUser, 8);
  }

  static isTeacher(): boolean {
    return this.checkPermission(UsersModule.currentUser, 4);
  }

  static isStudentHealth(): boolean {
    return this.checkPermission(UsersModule.currentUser, 2);
  }

  static isStudent(): boolean {
    return UsersModule.currentUser && (UsersModule.currentUser.role & 1) == 1;
  }
}
