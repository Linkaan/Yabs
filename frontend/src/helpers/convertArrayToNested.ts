export default function convertList(payload: any[], typeOfId: string): object { //eslint-disable-line
  return payload.reduce((acc, item) => {
    acc[item[typeOfId]] = item;
    return acc;
  }, {});
}
