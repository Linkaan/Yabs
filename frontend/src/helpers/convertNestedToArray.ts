export default function convertList(payload: any, typeOfId: string): any[] { //eslint-disable-line
  const res = Object.keys(payload).map((id) => payload[id]);
  return res;
}
