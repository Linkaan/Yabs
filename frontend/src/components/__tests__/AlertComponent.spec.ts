import { Wrapper } from '@vue/test-utils';
import { shallowFactory } from '@/helpers/testFactoryHelpers';
import Vue from 'vue';

import AlertComponent from '@/components/AlertComponent.vue';

describe('BaseModal.vue', () => {
  it('Renders corretly', async () => {
    const wrapper: Wrapper<Vue> = shallowFactory(AlertComponent);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
