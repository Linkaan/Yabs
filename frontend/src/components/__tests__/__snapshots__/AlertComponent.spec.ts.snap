// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertComponent.vue Renders corretly 1`] = `
<v-alert-stub color="" elevation="6" tag="div" transition="scale-transition" closelabel="$vuetify.close" dense="true" dismissible="true" icon="mdi-information" prominent="true" type="info" rounded="" class="onTop">

</v-alert-stub>
`;
